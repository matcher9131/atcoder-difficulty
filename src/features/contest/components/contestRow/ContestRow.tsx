import type { ReactNode } from "react";

export type ContestRowProps = {
    readonly headerCell: ReactNode;
    readonly problemCells: readonly ReactNode[];
};

export const ContestRow = ({ headerCell, problemCells }: ContestRowProps): ReactNode => {
    return (
        <tr className="contents">
            {headerCell}
            {problemCells}
        </tr>
    );
};
