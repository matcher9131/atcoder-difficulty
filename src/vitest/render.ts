import { render as originalRender, type RenderOptions } from "@testing-library/react";
import userEvent, { type UserEvent } from "@testing-library/user-event";
import type { ReactNode } from "react";

export const render = (
    element: ReactNode,
    options?: Omit<RenderOptions, "queries">,
): ReturnType<typeof originalRender> & { readonly user: UserEvent } => {
    return {
        user: userEvent.setup(),
        ...originalRender(element, options),
    };
};
