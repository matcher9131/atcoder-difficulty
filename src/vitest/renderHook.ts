import { renderHook as originalRenderHook } from "@testing-library/react";
import { Provider } from "jotai";

export const renderHook = <T, Props>(
    render: (initialProps: Props) => T,
): ReturnType<typeof originalRenderHook<T, Props>> => {
    return originalRenderHook(render, { wrapper: Provider });
};
