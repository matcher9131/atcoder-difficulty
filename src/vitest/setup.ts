import * as matchers from "@testing-library/jest-dom/matchers";
import type { TestingLibraryMatchers } from "@testing-library/jest-dom/matchers";
import { cleanup } from "@testing-library/react";
import { afterEach, expect } from "vitest";

// Extend Vitest's expect with jest-dom matchers
expect.extend(matchers);

// Merge type declaration of matchers of jest-dom to vitest
declare module "vitest" {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    interface Assertion<T = any> extends jest.Matchers<void, T>, TestingLibraryMatchers<T, void> {}
}

// Clean up the DOM after each test
afterEach(() => {
    cleanup();
});
