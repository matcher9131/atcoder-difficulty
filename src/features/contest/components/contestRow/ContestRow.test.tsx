import { describe, it, expect } from "vitest";
import { render, screen } from "@testing-library/react";
import { ContestRow } from "./ContestRow";

describe("ContestRow", () => {
    it("renders a <tr> with class 'contents'", () => {
        render(
            <table>
                <tbody>
                    <ContestRow
                        headerCell={<td data-testid="header-cell">Header</td>}
                        problemCells={[
                            <td key="1" data-testid="problem-cell-1">
                                Problem 1
                            </td>,
                            <td key="2" data-testid="problem-cell-2">
                                Problem 2
                            </td>,
                        ]}
                    />
                </tbody>
            </table>,
        );
        const tr = screen.getByRole("row");
        expect(tr).toHaveClass("contents");
    });

    it("renders the headerCell", () => {
        render(
            <table>
                <tbody>
                    <ContestRow headerCell={<td data-testid="header-cell">Header</td>} problemCells={[]} />
                </tbody>
            </table>,
        );
        expect(screen.getByTestId("header-cell")).toBeInTheDocument();
        expect(screen.getByTestId("header-cell")).toHaveTextContent("Header");
    });

    it("renders all problemCells", () => {
        render(
            <table>
                <tbody>
                    <ContestRow
                        headerCell={<td data-testid="header-cell">Header</td>}
                        problemCells={[
                            <td key="1" data-testid="problem-cell-1">
                                Problem 1
                            </td>,
                            <td key="2" data-testid="problem-cell-2">
                                Problem 2
                            </td>,
                        ]}
                    />
                </tbody>
            </table>,
        );
        expect(screen.getByTestId("problem-cell-1")).toBeInTheDocument();
        expect(screen.getByTestId("problem-cell-2")).toBeInTheDocument();
    });

    it("renders children in correct order", () => {
        render(
            <table>
                <tbody>
                    <ContestRow
                        headerCell={<td data-testid="header-cell">Header</td>}
                        problemCells={[
                            <td key="1" data-testid="problem-cell-1">
                                Problem 1
                            </td>,
                            <td key="2" data-testid="problem-cell-2">
                                Problem 2
                            </td>,
                        ]}
                    />
                </tbody>
            </table>,
        );
        const tr = screen.getByRole("row");
        const children = Array.from(tr.children);
        expect(children[0]).toHaveAttribute("data-testid", "header-cell");
        expect(children[1]).toHaveAttribute("data-testid", "problem-cell-1");
        expect(children[2]).toHaveAttribute("data-testid", "problem-cell-2");
    });

    it("matches snapshot", () => {
        const { container } = render(
            <table>
                <tbody>
                    <ContestRow
                        headerCell={<td>Header</td>}
                        problemCells={[<td key="1">Problem 1</td>, <td key="2">Problem 2</td>]}
                    />
                </tbody>
            </table>,
        );
        expect(container).toMatchSnapshot();
    });

    it("renders correctly with empty problemCells", () => {
        render(
            <table>
                <tbody>
                    <ContestRow headerCell={<td data-testid="header-cell">Header</td>} problemCells={[]} />
                </tbody>
            </table>,
        );
        expect(screen.getByTestId("header-cell")).toBeInTheDocument();
    });
});
