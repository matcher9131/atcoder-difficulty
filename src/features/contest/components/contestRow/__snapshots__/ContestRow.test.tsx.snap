// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ContestRow > matches snapshot 1`] = `
<div>
  <table>
    <tbody>
      <tr
        class="contents"
      >
        <td>
          Header
        </td>
        <td>
          Problem 1
        </td>
        <td>
          Problem 2
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
